### ISSUE TITLE
Case: output collapse on cross-lingual voice cloning — reproduces on the official PyTorch SDK

### ISSUE THREAD

**Issue by IIIIIllllIIIIIlllll** [breezeblue-ai/breeze-tts#14]
Salute to you all, and thank you for open-sourcing such an excellent model! Here are some issues I encountered while using an older GPU (2080Ti) — I hope you can help me with them!

**Environment** (the machine where the official SDK reproduced it):

- GPU: NVIDIA RTX 2080 Ti 22.5 GB (Turing, sm_75), driver595.71.05
- CPU: AMD Ryzen AI MAX+ 395; OS: Ubuntu 25.10, kernel 6.18.20
- Python 3.13.7; torch 2.9.1+cu126; transformers 4.57.3; accelerate 1.12.0
- Model: official Breeze-TTS-2 HF safetensors, `dtype: bfloat16`, eager attention
- Invocation: official repo's `infer.py`, defaults except the seed sweep below

**Important methodology note**: because the failure is a chaotic trajectory divergence, the failing seed is not stable across GPUs or PyTorch/CUDA versions — on a server GPU (sm_80/90) seed 51 may succeed while another seed fails. Please evaluate the **collapse rate over a seed sweep** (e.g. 20 seeds), not a single seed.

# Breeze-TTS-2 pronunciation-instability / output-collapse cases — attachments

Environment (all runs below): **official PyTorch SDK**, Breeze-TTS-2 HF
safetensors (bfloat16, eager attention), torch 2.9.1+cu126,
transformers 4.57.3, Python 3.13.7, Ubuntu 25.10, RTX 2080 Ti 22.5 GB,
driver 595.71.05. Generation settings all defaults (temperature 0.9,
top_k 50, top_p 1.0); only the seed varies.

## Case 1 — English voice cloning, the "audio.cpp" scenario (primary)

This reproduces the pronunciation instability reported downstream by the audio.cpp maintainer. His exact command text was not published; we used a close variant of his scenario (same "audio.cpp" token, same reference-audio family, 60 s reference).

Inputs:

- Mode: voice cloning
- Reference audio: ref-60.wav (included here), 60.0 s, 24 kHz — the
  "Mother Nature" sentence looped ~4.5 times (ASR-verified)
- Reference transcript: "Some call me nature. Others call me Mother Nature. I have been here for over four and a half billion years."
- Text: "Welcome to the audio.cpp plus BreezeTTS 2 clone validation. This request uses a long reference audio sample, so the speech encoder must preserve the speaker identity while keeping reference encoding bounded."
- Instruction: (none — scene default)

Note on the reference transcript (applies to both cases): the actual audio
continues with one more clause ("…twenty-two thousand five hundred times
longer than you."), so the transcript above is a truncated version of the
spoken content. This is the exact pairing used in all runs below (and in
the audio.cpp maintainer's original commands) — please keep it verbatim
when reproducing, since completing the transcript changes the token stream
and shifts which seeds collapse.

Official-SDK seed sweep on this exact input, seeds 42–51:

| Seed | Duration | ASR transcript of the "audio.cpp" region | Verdict |
|---|---|---|---|
| 42 | 16.8 s | "Welcome to the **audio. CPE** plus breeze TTS two…" | wrong pronunciation |
| 43 | 16.2 s | "Welcome to the **audio dot CPP** plus Breeze TTS two…" | clean |
| 44 | 18.0 s | "audio dot C P P **Finkenplus** breeze T T S two…" | gibberish insertion |
| 45 | 14.4 s | "audio. **Cpa** plus breeze tts two…" | wrong pronunciation |
| 46 | 33.9 s | "Welcome to the **Taramistars Varobiatu dot Coita, lawn of Phi Different…**" | **full gibberish collapse**, 2× duration |
| 47 | 14.9 s | "audio dot **CPO**…" | wrong pronunciation |
| 48 | 15.2 s | "audio. **CPT**…" | wrong pronunciation |
| 49 | 16.2 s | "audio dot **CPE**…" | wrong pronunciation |
| 50 | 15.5 s | "audio dot C P P plus breeze T T S two…" | clean |
| 51 | 17.7 s | "audio dot C P P **P fast** breeze T T E S two…" | gibberish insertion |

Summary: only 2/10 seeds render the sentence cleanly; 5/10 mispronounce
"audio.cpp"; 2/10 insert gibberish words; 1/10 collapses entirely into
33.9 s of nonsense. We concede "audio.cpp" is a hard literal token — but
seed 46's wholesale collapse cannot be explained by that.

Attached renders (official SDK):

| File | Seed | Note |
|---|---|---|
| official-sdk-en-60sref-seed46-GIBBERISH.wav | 46 | full collapse, 33.9 s nonsense |
| official-sdk-en-60sref-seed44-gibberish-insert.wav | 44 | "Finkenplus" insertion |
| official-sdk-en-60sref-seed51-gibberish-insert.wav | 51 | "P fast" insertion |
| official-sdk-en-60sref-seed42-audioCPE.wav | 42 | "audio CPE" — the variant flagged as a regression downstream; the official SDK produces it too |
| official-sdk-en-60sref-seed43-normal.wav | 43 | clean contrast |
| ref-60.wav | — | reference audio input |

## Case 2 — Chinese voice cloning, runaway gibberish (supplementary)

Same failure family on an ordinary Chinese sentence — no adversarial
tokens involved.

Inputs:

- Reference audio: b.wav (included here), English male, ~14 s
- Reference transcript: same as above
- Text: 它必须具有这种能力，因为没有任何人为的力量，可以足够迅速或堪以适当地，为它作出调整和修正。
- Instruction: 软软糯糯的少女音，表达轻快而富有感染力

Official-SDK seed sweep, seeds 42–61: **3/20 collapsed** (seeds 51, 57, 58)
into 16.7–35.5 s of full-volume English gibberish instead of ~10 s of
Chinese; 17/20 normal (9.3–11.0 s, correct Chinese per ASR).

| File | Seed | Duration | Content |
|---|---|---|---|
| official-sdk-clone-seed51-COLLAPSE.wav | 51 | 27.6 s | English gibberish |
| official-sdk-clone-seed57-COLLAPSE.wav | 57 | 16.7 s | English gibberish |
| official-sdk-clone-seed58-COLLAPSE.wav | 58 | 35.5 s | English gibberish |
| official-sdk-clone-seed42-normal.wav | 42 | 10.5 s | correct Chinese (contrast) |
| b.wav | — | 14 s | reference audio input |

## Reproduction notes

- Collapse/instability rate on our hardware: Case 1 ≈ 80% imperfect
  (10% full collapse), Case 2 ≈ 15% full collapse, per seed.
- The failing seeds are **not stable across GPUs/toolchains** — sampling is
  chaotic, so on your machines these specific seeds may pass while others
  fail. Please evaluate over a seed sweep (e.g. 42–61), not a single seed;
  ideally on both newer server GPUs and older consumer ones (ours is
  sm_75).
- All transcripts above were produced by an independent ASR model
  (Qwen3-ASR-0.6B) from the official SDK's own rendered wav files.

[breeze-tts2-collapse-cases.zip](https://github.com/user-attachments/files/31844881/breeze-tts2-collapse-cases.zip)

**Comment by IIIIIllllIIIIIlllll** [comment]
# Breeze-TTS-2: Reproducing Two Defects with the Official SDK

## How we tested

- Model: Breeze-TTS-2, official PyTorch weights (eager loading)
- Environment: RTX 2080 Ti, official SDK `infer.py`, all generation settings at defaults (CFG 1.0, repetition penalty 1.1, max 1500 new tokens)
- Method: **pure voice design, no reference audio**. Same input rendered seed by seed, seeds 42–67 (26 renders)
- Verification: Qwen3-ASR-0.6B transcription diffed against the input text; per-band RMS tail analysis (LF <300 Hz / MF 300–3000 Hz / HF 3k–10k Hz); human listening review of all flagged samples

Reproduction command (single render):

```bash
python infer.py <model_dir> \
  --text '哎呀呀~ 杂鱼杂鱼
你以为你这样就能赢我吗？杂鱼
哼，别用那种眼神看着我啦，小杂鱼
明明就是个杂鱼，还敢来挑战本小姐，真是勇气可嘉呢，杂鱼
喂，杂鱼，快把谢谢大小姐说给我听听，不然我就把你按在地上摩擦哦杂鱼' \
  --instruction '10岁萝莉音，小恶魔，傲娇奶凶，软糯，语调上扬，尾音拖长，语速轻快，充满挑衅和得意的嘲讽感。' \
  --output offseed-56.wav --seed 56
```

(Design instruction, translated: "10-year-old loli voice, little devil, tsundere bratty, soft and squishy, rising intonation, dragged-out line endings, brisk pace, full of provocative and smug mockery.")

## Finding 1: seed 56 — audible tail collapse

The final sentence ("喂，杂鱼，快把谢谢大小姐……") fades progressively, as if the speaker is walking away.

- The mid band (300–3000 Hz — the speech presence range the ear is most sensitive to) drops **7.5 dB** at the tail vs the body, with the high band down 2.6 dB in parallel: a full-spectrum fade
- For contrast: the other 25 renders only lose low-frequency phonation energy at the end (the designed "dragged soft tail" the instruction asks for), with MF/HF flat — those sound normal
- Audio: `official-seed56-tail-collapse.wav`

## Finding 2: seed 67 — mispronounced characters

- The opening "哎呀呀" (āi yā yā) is read as "**爷爷**" (yé ye)
- The second line's "杂鱼" (zá yú) is read as "**大鱼**" (dà yú)
- Confirmed by both ASR transcription and human listening — the syllables are far apart, so this is not a recognition artifact
- Audio: `official-seed67-mispronunciation.wav` (this render's tail is normal; only the reading is wrong)

## Conclusion

Both defects — tail volume collapse and character misreads — are produced directly by the **official PyTorch SDK at default settings**. They are model-intrinsic sampling trajectory degeneration, not artifacts of any downstream inference implementation. Sampling is sensitive to sub-ulp logit noise, so which seed fails varies by machine/toolchain: **retrying a single seed may not reproduce the issue — a seed sweep is required** (in this 26-seed sweep, each defect appeared once).

[官方完整-seed56-中高频双降.wav](https://github.com/user-attachments/files/31849013/-seed56-.wav)
[官方完整-seed67-尾渐弱.wav](https://github.com/user-attachments/files/31849015/-seed67-.wav)
[官方完整-seed64-中高频双降.wav](https://github.com/user-attachments/files/31849014/-seed64-.wav)

**Comment by JiaxinTian1** [comment]
Thanks for the detailed report. 

We suspect this may be related to the RTX 2080 Ti’s Turing (sm_75) kernel path, which lacks native BF16 support. The resulting fallback kernels and numerical differences may be amplified by sampling and lead to different generation trajectories. We recommend testing on an Ampere-or-newer GPU with native BF16 support, or explicitly using FP16/FP32 on the 2080 Ti. For comparison, we have attached our H100 BF16 outputs generated with the same prompt and seed.

[local-h100-seed67-mispronunciation-check.wav](https://github.com/user-attachments/files/31953942/local-h100-seed67-mispronunciation-check.wav)

We also found two input-related issues. In Case 1, the 60-second reference repeats the same sentence several times while the transcript is incomplete, so the reference audio and text are not aligned. Please use clean, non-looping audio with an exact transcript of everything spoken. In Case 2, a reference voice is combined with a conflicting voice instruction, so this is Voice Direction rather than Voice Clone.

PR #17 fixes an additional routing bug discovered during this investigation: reference-only requests were previously given a default instruction. After the fix, reference audio without an instruction uses pure Voice Clone; providing a non-empty instruction intentionally selects Voice Direction. For Voice Clone, please omit instruction entirely and use cfg_scale=1.

Please retry on the latest main after PR #17 is merged, using a supported dtype and correctly aligned reference input. If it still collapses, please share the exact dtype, seed sweep, inputs, and outputs from that corrected setup.

**Comment by IIIIIllllIIIIIlllll** [comment]
> Thanks for the detailed report.
> 
> We suspect this may be related to the RTX 2080 Ti’s Turing (sm_75) kernel path, which lacks native BF16 support. The resulting fallback kernels and numerical differences may be amplified by sampling and lead to different generation trajectories. We recommend testing on an Ampere-or-newer GPU with native BF16 support, or explicitly using FP16/FP32 on the 2080 Ti. For comparison, we have attached our H100 BF16 outputs generated with the same prompt and seed.
> 
> [local-h100-seed67-mispronunciation-check.wav](https://github.com/user-attachments/files/31953942/local-h100-seed67-mispronunciation-check.wav)
> 
> We also found two input-related issues. In Case 1, the 60-second reference repeats the same sentence several times while the transcript is incomplete, so the reference audio and text are not aligned. Please use clean, non-looping audio with an exact transcript of everything spoken. In Case 2, a reference voice is combined with a conflicting voice instruction, so this is Voice Direction rather than Voice Clone.
> 
> PR [#17](https://github.com/breezeblue-ai/breeze-tts/pull/17) fixes an additional routing bug discovered during this investigation: reference-only requests were previously given a default instruction. After the fix, reference audio without an instruction uses pure Voice Clone; providing a non-empty instruction intentionally selects Voice Direction. For Voice Clone, please omit instruction entirely and use cfg_scale=1.
> 
> Please retry on the latest main after PR [#17](https://github.com/breezeblue-ai/breeze-tts/pull/17) is merged, using a supported dtype and correctly aligned reference input. If it still collapses, please share the exact dtype, seed sweep, inputs, and outputs from that corrected setup.

Understood. Thank you for taking the time to reply despite your busy schedules.
